ZepMemory implementation for LlamaIndex

bashiryounis opened the pull request:
# Add ZepMemory Integration for LlamaIndex

## Overview
This PR introduces a `ZepMemory` class that provides integration between LlamaIndex and Zep, allowing for persistent memory, context enrichment, and retrieval capabilities in LlamaIndex applications.

## Features
- Bidirectional sync between LlamaIndex's `ChatMemoryBuffer` and Zep's memory store
- Support for session-based memory management
- Context enrichment with facts, summaries, and relevant information from previous sessions
- Memory search capabilities across sessions for user-specific context
- Message truncation to prevent oversized payloads
- Graceful error handling for resilient operation

## Implementation Details
- Extends LlamaIndex's `BaseMemory` class
- Handles message format conversion between LlamaIndex and Zep
- Preserves metadata across systems
- Supports system, user, assistant, and tool message roles
- Implements memory lifecycle methods (get, put, set, reset, search)

## Dependencies
- Requires `zep-python` client library

logan-markewich commented on llama-index-integrations/memory/llama-index-memory-mem0/llama_index/memory/zep/base.py:
```diff
@@ -0,0 +1,248 @@
+        user_id: Optional[str] = None,
+        memory_key: str = "chat_history",
+        max_message_length: int = 2500,
+        **kwargs,
```
no need for kwargs if we aren't using them

bashiryounis replied:
I agree, I'll remove `**kwargs` from the signature to simplify the code. Thanks for the suggestion!

logan-markewich commented on llama-index-integrations/memory/llama-index-memory-mem0/llama_index/memory/zep/base.py:
```diff
@@ -0,0 +1,248 @@
+            memory_key=memory_key,
+            max_message_length=max_message_length,
+        )
+        self._client = zep_client
```
does zep have an async client? There are a few async methods in the base class that would be nice to implement

bashiryounis replied:
I’ve updated the ZepMemory implementation to include full asynchronous support. The new code now implements async versions for all key operations (e.g. `_async_sync_from_zep`, `aget`, `aput`, `aset`, `areset`, and `asearch`). When using an `AsyncZep` client with the `is_async` flag set to `True`, the memory will use these `async` methods; otherwise, it falls back to the synchronous methods.

Let me know if you have any further suggestions!

bashiryounis replied:
@logan-markewich  could you please let  me know if you have any further suggestions

bashiryounis commented on llama-index-integrations/memory/llama-index-memory-mem0/llama_index/memory/zep/base.py:
Thank you for the feedback. You’re absolutely right—it should be its own integration. I apologize for the oversight in placing it within the mem0 folder. I'll make the necessary adjustments promptly.

logan-markewich commented:
To add an integration, you'll need more than a single file -- check out other integrations for examples, but you'll need a README and a pyproject.toml at a minimum 

Tests would also be nice

bashiryounis commented:
- Added a `README` and a `notebook` to demonstrate usage and integration with the Zep memory server for testing and exploration.

- Added tests for both `sync` and `async` functionality.

- Added a `pyproject.toml ` file for  project configuration .

Feel free to check it out — I’d love to hear your thoughts and feedback!

Thanks! 🙌

bashiryounis commented:
Hi @logan-markewich,
Thanks again for your feedback earlier 🙌
Please let me know if there’s anything else needed to finalize this contribution.

Appreciate your time and guidance!

Thanks,
